Feat/#31 캘린더 관련 기능 구현

yujeongkwon opened the pull request:
캘린더 관련 기능 구현했습니다.
MVP로 인해 현재 요구사항과 안맞아 에러가 나는 부분은 일단 임시로 수정하거나 주석 처리해두었습니다.

yujeongkwon committed:
init : 캘린더 구현을 위한 초기 세팅

- Profile 클래스 생성
- User 클래스 RDB 기반 임시 생성
- 위에 따른 식별자였던 nickname을 loginId로 사용하도록 급한 것만 변경

yujeongkwon committed:
chore : 도메인별 클래스 패키지 정리

yujeongkwon committed:
chore : image 관련 도메인 분리
        및 user의 식별자 nickname에서 loginId로 추가 변경

yujeongkwon committed:
feat(Calendar): 캘린더 생성

- CalendarMember 클래스 생성
- 캘린더 프로필 이미지 저장 시, "profile/" 경로로 저장할 수 있도록 구현
- S3Config 클래스의 getDefaultImage()에서 현재는 임시 url로 설정함, 추후 기본 이미지 url이 결정되면 수정필요

yujeongkwon committed:
feat(Calendar): 캘린더 조회

- 기본 이미지 가져오는 것을 상수 클래스로 이동

yujeongkwon committed:
feat(Calendar): 캘린더 수정 및 이미지 관련 패키지 이동

yujeongkwon committed:
feat(Calendar): 캘린더 나가기

delete(Profile): 프로필 엔티티 제거 및 관련 속성 CalendarMember에 추가

yujeongkwon committed:
feat(Calendar): 캘린더 사용자 초대

yujeongkwon committed:
feat(Calendar): 캘린더 멤버 목록 조회

yujeongkwon committed:
feat(Calendar): 캘린더 멤버 초대 사용자 검색 목록 구현

yujeongkwon committed:
refactor(Calendar): 로직 개선 및 사용자 검색 목록 성능 최적화

- 로그인 사용자가 해당 캘린더 멤버인지 확인하는 로직 메서드에 모두 추가
- 캘린더 사용자 검색 목록 성능 최적화

yujeongkwon committed:
feat(Calendar): 사용자가 속한 캘린더 목록 불러오기 구현

yujeongkwon committed:
test(Calendar): 캘린더 서비스 통합 테스트

yujeongkwon commented:
빌드에 성공했습니다.😊

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/domain/Calendar.java:
```diff
@@ -0,0 +1,86 @@
+    private String name;
+
+    @Column(name = "profile_image_url")
+    private URL profileImageUrl;
```
캘린더 멤버별 프로필 이미지가 추가될 것을 고려해 calendarImageUrl이나 calendarProfileImageUrl과 같이 명시해주면 어떨까요??

-> 라고 생각했는데 코드를 읽다보니 캘린더나 캘린더 멤버 모두 객체 이름 뒤에 쓰면 굳이 변수 이름에 명시해주지 않아도 괜찮을 것 같네요!

yujeongkwon replied:
ㅎㅎ 굿👍🏻

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/dto/response/CalendarMemberResponseDto.java:
```diff
@@ -0,0 +1,32 @@
+    private Integer userId;
+    private String name;
+    private URL profileImageUrl;
+    private String status;  //false, invited, accepted
```
invited나 accepted는 의미가 잘 나타나있는 것 같은데 false 대신 uninvited로 쓰는건 어떻게 생각하세요? 😀

yujeongkwon replied:
수정하겠습니다!

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/domain/Calendar.java:
```diff
@@ -0,0 +1,86 @@
+        member.setCalendar(this);
+    }
+
+    private void validateName(String name) {
```
유효성 검사도 도메인에 넣은것이 인상적이네요! 객체가 자기의 일을 할 수 있도록 책임을 잘 분리한 것 같아요. 저의 코드에도 적용해봐야겠습니당 👍🏻

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/dto/response/RegisterCalendarResponseDto.java:
```diff
@@ -0,0 +1,16 @@
+
+@Getter
+public class RegisterCalendarResponseDto {
+    private int calendarId;
```
다른 곳에선 캘린더 아이디를 Integer로 하셨던데, 여기만 특별히 int로 지정한 이유가 있나요!

yujeongkwon replied:
Dto에서 Id 같은 경우는 무조건 들어간느 거라 int나 Integer 아무거나 상관없다고 생각하여 편히 구현하다가 이것 저것 섞인 것 같습니다. Dto에서 Id 값들은 모두 필수로 들어가야하므로 int로 다 통일성을 주는게 가독성 측면에서도 효율적인 면에서도 좋을 것 같습니다. 
지적해주셔서 감사해요!

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/dto/response/CalendarSummaryResponseDto.java:
```diff
@@ -0,0 +1,22 @@
+    private final Integer calendarId;
+    private final String name;
+    private final String profileImageUrl;
+    private final Integer participantsCount;
```
여기 responseDto에만 멤버들을 final로 선언한 이유가 있나용?

yujeongkwon replied:
보통 dto는 값의 전달만을 담당하기 때문에 불변성을 위해 final을 쓰는게 좋다고 생각했는데 그 생각을 해당 Dto에서만 했나보네요 ㅎ.. final을 습관적으로 자주써야겠다고 생각했는데 지적해주셔서 감사합니다. 다른 Dto에도 적용해야겠어용

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/service/CalendarService.java:
```diff
@@ -0,0 +1,177 @@
+        Calendar calendar = saveCalendar(name, profileImageUrl);
+        saveCalendarMember(calendar, user);
+
+        return new RegisterCalendarResponseDto(calendar.getId());
```
캘린더 생성하고 RegisterCalendarResponseDto 를 반환하는 것은 캘린더를 새로 만들고 나면 메인페이지로 로드돼서 바로 캘린더 요약 정보를 넘겨주는 건가용?

yujeongkwon replied:
넵! 맞습니다

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/service/CalendarService.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    @Transactional(readOnly = true)
+    public CalendarResponseDto findCalendarById(int calendarId, String loginId) {
```
여기도 캘린더 아이디를 int로 쓴 이유가 있을까요! 다른 곳에는 Integer로 쓰신 것 같아서요 😲😀

yujeongkwon replied:
위에서 설명한 것과 같습니다 ㅎ.. 다른 메서드들도 다 int로 바꿀 예정입니다! 혼란을 줘서 죄송해요😂

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/service/CalendarService.java:
```diff
@@ -0,0 +1,177 @@
+        calendarMemberRepository.save(calendarMember);
+    }
+
+    private void changeProfileImage(MultipartFile newProfileImage, Calendar calendar) {
```
이 메서드는 캘린더 도메인에 있는 changeProfileImage 메서드랑 어떻게 다른건가용??

yujeongkwon replied:
해당 메서드는 이미지를 변화시킨다면 aws S3에 업로드하고 업로드된 경로의 URL을 가져오고, 도메인의 changeProfileImage()는 가져온 URL을 본래 객체 프로필이미지 URL을 변경하는 것입니다!

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/service/CalendarService.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    @Transactional(readOnly = true)
+    public List<CalendarSummaryResponseDto> getCalendarsByUser(Integer userId) {
```
userId로 캘린더 가져오는 것 같은데 getCalendarsByUserId 말고 getCalendarsByUser로 하신 이유가 있나용? 😀

yujeongkwon replied:
사용자가 속한 캘린더 목록을 가져오는 거라, 사용자의 아이디로 캘린더들을 가져온다는 구현적 느낌보다 사용자의 캘린더 목록이라는 비즈니스적 의미에 더 가깝게 하고 싶었습니다!

o-jeong commented on src/main/java/com/fluffy/SharingCalendar/calendar/service/CalendarService.java:
```diff
@@ -0,0 +1,177 @@
+                .orElseThrow(() -> new CustomException(CALENDAR_MEMBER_NOT_FOUND));
+    }
+
+    private void checkDuplicateInvitation(Integer calendarId, Integer userId) {
```
checkUserIncluded 메서드명처럼 check+대상+상태로 네이밍을 한다면 이 메서드 명도 checkInvitationDuplicated로 통일하는 것도 좋을 것 같아요! 혹시 유정님의 생각은 어떠신가요?ㅎㅎ

yujeongkwon replied:
좋은 것 같습니다! 항상 메서드 네이밍하는 게 가장 어려웠는데 ㅠ 역쉬 영잘알이십니다👍🏻

yujeongkwon committed:
refactor(Calendar): 코드 리뷰 수정 반영

- Dto와 캘린더 서비스 매개변수의 Integer를 int로 변환
- 메서드명 재명명
- 캘린더 멤버 검색 시, 캘린더와 아무 관련이 없는 멤버 상태를 false->uninvited로 변경

yujeongkwon reviewed (commented):
자세한 코드리뷰 감사합니다. 코드 리뷰 반영하여 수정하였습니다!

yujeongkwon commented:
빌드에 성공했습니다.😊
